python/ray/utils.py

```diff
@@ -215,3 +217,32 @@ def merge_dicts(d1, d2):
     d = d1.copy()
     d.update(d2)
     return d
+
+
+class ThreadSafeProxy(object):
+
+    def __init__(self, orig_obj, lock):
+        self.orig_obj = orig_obj
+        self.lock = lock
+        self._wrapper_cache = {}
+
+    def __getattr__(self, attr):
+        orig_attr = getattr(self.orig_obj, attr)
+        if callable(orig_attr):
+            wrapper = self._wrapper_cache.get(attr)
+            if wrapper is None:
+                @functools.wraps(orig_attr)
+                def _wrapper(*args, **kwargs):
+                    with self.lock:
+                        return orig_attr(*args, **kwargs)
+                self._wrapper_cache[attr] = wrapper = _wrapper
+            return wrapper
+        else:
+            return orig_attr
+
+
+def thread_safe_client(client, lock=None):
```

Comment:
You mean letting the users use `ThreadSafeProxy` directly?

Reply:
yeah

Reply:
The reason why I wanted to do this is because `ThreadSafeProxy` is the implementation detail, doesn't need to be exposed to the users.
The users only need a helper function that can make their clients synchronized.
Doing this also enable us